src/thunderbots/software/ai/world/team.cpp

```diff
@@ -1,33 +1,145 @@
 #include "team.h"
+#include <set>
+#include "shared/constants.h"
 
-Team::Team()
+Team::Team(const std::chrono::milliseconds robot_expiry_buffer_milliseconds)
+    : team_robots(),
+      goalie_id(),
+      robot_expiry_buffer_milliseconds(robot_expiry_buffer_milliseconds)
 {
 }
 
-void Team::update(std::vector<Robot> &team_robots)
+void Team::updateRobots(const std::vector<Robot>& new_robots,
+                        const std::chrono::steady_clock::time_point timestamp)
 {
+    // Update the robots, checking that there are no duplicate IDs in the given data
+    std::set<unsigned int> robot_ids;
+    for (const auto& robot : new_robots)
+    {
+        auto duplicate_id = robot_ids.insert(robot.id());
+        if (duplicate_id.second)
+        {
+            // TODO: Multiple robots on the same team with the same id. Throw exception
+            // See https://github.com/UBC-Thunderbots/Software/issues/16
+        }
+
+        auto it = team_robots.find(robot.id());
+        if (it != team_robots.end())
+        {
+            // The robot already exists on the team. Find and update the robot
+            team_robots.at(robot.id()).updateState(robot);
+        }
+        else
+        {
+            // This robot does not exist as part of the team yet. Add the new robot
+            team_robots.insert(std::make_pair(robot.id(), robot));
+        }
+    }
+
+    removeExpiredRobots(timestamp);
+}
+
+void Team::updateState(const Team& new_team_data,
+                       const std::chrono::steady_clock::time_point timestamp)
+{
+    updateRobots(new_team_data.getAllRobots(), timestamp);
+    this->goalie_id = new_team_data.goalie_id;
+}
+
+void Team::updateStateToPredictedState(
+    const std::chrono::steady_clock::time_point timestamp)
+{
+    // Update the state of all robots to their predicted state
+    for (auto it = team_robots.begin(); it != team_robots.end(); it++)
+    {
+        it->second.updateStateToPredictedState(timestamp);
+    }
 }
 
-void Team::update(const thunderbots_msgs::Team &team_msg)
+void Team::removeExpiredRobots(const std::chrono::steady_clock::time_point timestamp)
 {
+    // Check to see if any Robots have "expired". If it more time than the expiry_buffer
+    // has passed, then remove the robot from the team
+    for (auto it = team_robots.begin(); it != team_robots.end();)
+    {
+        if (std::chrono::duration(timestamp - it->second.lastUpdateTimestamp()) >
+            robot_expiry_buffer_milliseconds)
+        {
+            team_robots.erase(it++);
+        }
+        else
+        {
+            it++;
+        }
+    }
+}
+
+void Team::assignGoalie(unsigned int new_goalie_id)
+{
+    goalie_id = new_goalie_id;
```

Comment:
Should this throw an exception if a robot with the given goalie id does not exist on the team?